content/200-concepts/100-components/02-prisma-client/052-advanced-type-safety/index.mdx

```diff
@@ -0,0 +1,182 @@
+
+The following example shows two files duplicating query options.
+
+```ts file=find-users.js
+import { PrismaClient } from '@prisma/client'
+
+const prisma = new PrismaClient()
+
+// Returns all users `name` and `email` but omits their `posts`
+const users = await prisma.user.findMany({
+  select: {
+    name: true,
+    email: true,
+    posts: false,
+  },
+})
+```
+
+```ts file=find-unique-user.js
+import { PrismaClient } from '@prisma/client'
+
+const prisma = new PrismaClient()
+
+// Returns a single users `name` and `email` but omits their `posts`
+const user = await prisma.user.findUnique({
+  where: { id: 1 }
+  select: {
+    name: true,
+    email: true,
+    posts: false,
+  }
+})
+```
+
+Notice both files use the same `select` options. Although there is nothing wrong with this approach, this code could be streamlined and the duplicated code removed.
+
+The following uses the `validator` and the generated type `UserSelect` to create a reusable `select` query option.
+
+```ts file=my-custom-query
+import { Prisma, PrismaClient } from '@prisma/client'
+
+export const selectNameEmailNotPosts = Prisma.validator<Prisma.UserSelect>()({
+  name: true,
+  email: true,
+  posts: false,
+})
+```
+
+This new query option can now be used in both of the above queries, its type-safe and it reduces code duplication in your codebase.
+
+```ts
+import { PrismaClient } from '@prisma/client'
+import { selectNameEmailNotPosts } from './my-custom-query'
+
+const prisma = new PrismaClient()
+
+const users = await prisma.user.findMany({
+  select: selectNameEmailNotPosts,
+})
+
+const user = await prisma.user.findUnique({
+  where: { id: 1 }
+  select: selectNameEmailNotPosts
+})
+```
+
+### Combining Prisma.validator with form input
+
+The following example creates a function from the `Prisma.validator` which can be used when interacting with user created data, such as form inputs.
+
+```ts
+import { Prisma, PrismaClient } from '@prisma/client'
+
+const prisma = new PrismaClient()
+
+// Create a new function and pass the parameters onto the validator
+const createUserAndPost = (name: string, email: string, postTitle: string, profileBio: string) => {
+  return Prisma.validator<Prisma.UserCreateInput>()({
+    name,
+    email,
+    posts: {
+      create: {
+        title: postTitle,
+      },
+    },
+    profile: {
+      create: {
+        bio: profileBio,
+      },
+    },
+  })
+}
+
+const findSpecificUser = (name: string, email: string) => {
+  return Prisma.validator<Prisma.UserWhereInput>()({
+    name,
+    email,
+  })
+}
+
+// Create the user in the database based on form input
+await prisma.user.create({
+  data: createUserAndPost('Rich', 'rich@boop.com', 'Life of Pie', 'Learning each day'),
+})
+
+// Find the specific user based on form input
+const oneUser = await prisma.user.findUnique({ where: findSpecificUser('Rich', 'rich@boop.com') })
```

Comment:
using `findUnique` with two parameters is probably not the best example as the point of `findUnique` is to find a record by a single unique property (so the `email` in this case would be enough already)